File: 2018/day06.php
<?php

function manhattanDistance(array $xy1, array $xy2) : int
{
    return abs($xy1[0] - $xy2[0]) + abs($xy1[1] - $xy2[1]);
}

$coords = array_filter(array_map(function(string $line) {
    return (empty($line)) ? null : array_map('intval', explode(',', $line));
}, explode("\n", file_get_contents(__DIR__.'/inputs/06.txt'))));

$maxX = array_reduce($coords, function($acc, $item) {
    return max($acc, $item[0]);
}, 0);

$maxY = array_reduce($coords, function($acc, $item) {
    return max($acc, $item[1]);
}, 0);

// which grid points are owned by what coord
$gridOwnership = [];

for ($y = 0; $y <= $maxY; $y++) {
    for ($x = 0; $x <= $maxX; $x++) {
        $pointDistances = [];

        foreach ($coords as $k => $point) {
            $distance = manhattanDistance([ $x, $y ], $point);

            if (!array_key_exists($distance, $pointDistances)) {
                $pointDistances[$distance] = [];
            }

            $pointDistances[$distance][] = $k;
        }

        $minDist = min(array_keys($pointDistances));
        if (count($pointDistances[$minDist]) > 1) {
            continue;
        }

        $gridOwnership[$x][$y] = $pointDistances[$minDist][0];
    }
}

$areas = array_fill(0, 50, 0);
$infiniteAreas = [];

for ($y = 0; $y <= $maxY; $y++) {
    for ($x = 0; $x <= $maxX; $x++) {

        if (isset($gridOwnership[$x][$y])) {
            $coordKey = $gridOwnership[$x][$y];
            $areas[$coordKey]++;

            if ($x === 0 || $y === 0 || $x === $maxX || $y === $maxY) {
                $infiniteAreas[] = $coordKey;
            }
        }
    }
}

$infiniteAreas = array_unique($infiniteAreas);

arsort($areas, SORT_NUMERIC);

foreach ($areas as $coordKey => $area) {
    if (!in_array($coordKey, $infiniteAreas)) {
        var_dump($coords[$coordKey], $area);
        break;
    }
}

File: 2018/day06.py
from collections import defaultdict

f = open('inputs/06.txt', 'r')
coords = set([ (int(line.split(',')[0]), int(line.split(',')[1])) for line in f.read().splitlines() ])

def manhattan(c1, c2):
    return abs(c1[0] - c2[0]) + abs(c1[1] - c2[1]);

max_x = max(coords, key=lambda v: v[0])[0]
max_y = max(coords, key=lambda v: v[1])[1]

grid = {}
part_2 = 0

for y in range(0, max_y):
    for x in range(0, max_x):
        distances = { coord: manhattan(coord, (x, y)) for coord in coords }
        min_distance = min([ v for v in distances.values() ])
        min_distance_coords = [ coord for coord in distances if distances[coord] == min_distance ]

        if len(min_distance_coords) == 1:
            grid[(x, y)] = min_distance_coords[0]

        if sum(distances.values()) < 10000:
            part_2 += 1

infinites = set()
areas = defaultdict(int)

for xy in grid:
    if xy[0] == 0 or xy[0] == max_x or xy[1] == 0 or xy[1] == max_y:
        infinites.add(grid[xy])

    if grid[xy] not in infinites:
        areas[grid[xy]] += 1

part_1 = max(set(areas.values()) - infinites)

print(part_1, part_2)
